### PR TITLE
Use a reduction implementation for unique when dtype is bool on CPU

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109695

### CODE REVIEW

File: aten/src/ATen/native/Unique.cpp [hunk 1]
```diff
@@ -30,6 +30,74 @@ namespace native{
...
 
+// This unique implementation when dtype is bool is mapped
+// from UniqueCub.cu which uses a reduction to find the number of
+// true values.
+std::tuple<Tensor, Tensor, Tensor> unique_cpu_bool_template(
+    const Tensor& self,
+    const bool return_inverse,
+    const bool return_counts) {
+  const Tensor& input = self.contiguous();
+  bool* input_data = input.data_ptr<bool>();
```
Comment (lezcano) [inline aten/src/ATen/native/Unique.cpp:41]: Does this work when the input is empty, or should we have a escape hatch in the function that calls into this one?

**Review by lezcano — APPROVED** [review]
Nice. Just one comment.

File: aten/src/ATen/native/Unique.cpp (hunk 1 above)
Comment (mingfeima) [reply aten/src/ATen/native/Unique.cpp:41]: Oh thanks for the remainder ! No this won't work properly when input is empty, it is fixed now!

**Review by lezcano — APPROVED** [review]
Could you add a tiny test case with an empty tensor to the opinfo or even a standalone test to be on the safe side? Otherwise LGTM

**Comment by mingfeima** [comment]
> Could you add a tiny test case with an empty tensor to the opinfo or even a standalone test to be on the safe side? Otherwise LGTM

Oh I just did some debug, the empty input is already covered in https://github.com/pytorch/pytorch/blob/main/test/test_sort_and_select.py#L818 and even without this code
```C++
if (numel == 0) {
  if (return_inverse) {
    inverse_indices.resize_(input.sizes());
  }
  return std::make_tuple(output, inverse_indices, counts);
 }
```
the old patch is good enough to pass this empty input test, I thought `inverse_indices` would not be properly resized so i said it was going to fail but it turns out `invese_indices` will always be resized to the same size of `self`. But anyway i feel more comfortable to add an explicit path to handle the empty input. Thanks for the remainder!

**Pull request closed without merging**